source-container-build/app/source_build.py

```diff
@@ -720,6 +765,126 @@ def index(self) -> Index:
+        """Check if an archive member is a link of RPM source
+
+        Example arcname: ./rpm_dir/foo-1.0.src.rpm
+        """
+        dirname, basename = os.path.split(member.name)
+        return member.issym() and dirname == "./rpm_dir" and basename.endswith(".src.rpm")
+
+    def _is_blob_file(self, member: TarInfo) -> bool:
+        """Check if an archive member is a blob file"""
+        regex: Final = r"\./blobs/sha256/[0-9a-f]+"
+        return member.isreg() and re.fullmatch(regex, member.name) is not None
+
+    def _extract(self) -> None:
+        """Extract symlink and blob members"""
+        with tarfile.open(self._layer.path, "r") as tar:
+            for member in tar:
+                if self.is_rpm_src(member):
+                    self._symlink_member = member
+                elif self.is_extra_src(member):
+                    self._symlink_member = member
+                    fo = tar.extractfile(member)
+                    try:
+                        with tarfile.open(fileobj=fo, mode="r") as extra_src_tar:
+                            files = [m for m in extra_src_tar.getmembers() if m.isreg()]
+                            self._extra_source = files[0]
+                    finally:
+                        fo.close()  # type: ignore
+                elif self._is_blob_file(member):
+                    self._blob_member = member
+
+    def _check(self) -> None:
+        """Check if expected members are found
+
+        A layer generated by BSI must have symlink and blob members, and the
+        symlink member links to the blob one.
+        """
+        err_prefix = "Invalid layer generated by BuildSourceImage."
+        if self.symlink_member is None:
+            raise ValueError(f"{err_prefix} No symlink member is found.")
+        if self.is_extra_src(self.symlink_member) and self.extra_source is None:
+            raise ValueError(f"{err_prefix} Missing extra source.")
+        if self.blob_member is None:
+            raise ValueError(f"{err_prefix} No blob member is found.")
+
+        dir_name, _ = os.path.split(self.symlink_member.name)
+        normalized_link_path = os.path.normpath(
+            os.path.join(dir_name, self.symlink_member.linkname)
+        )
+        if normalized_link_path != os.path.normpath(self.blob_member.name):
+            raise ValueError(
+                f"{err_prefix} Symlink {self.symlink_member.name} does not link to the blob."
+            )
+
+    def hash_key(self):
+        if self.extra_source:
+            artifact_name = self.extra_source.name
+        else:
+            artifact_name = self.symlink_member.name
+        return self.blob_member.name, artifact_name
+
+    def __hash__(self):
+        return hash(self.hash_key())
+
+    def __eq__(self, other: object) -> bool:
+        """Check if this layer contains same content of the other"""
+        if not isinstance(other, BSILayer):
+            return False
+        return hash(self) == hash(other)
```

Comment:
```suggestion
        return self.hash_key() == other.hash_key()
```
Otherwise different objects could be considered equal due to hash collisions

Reply:
Updated.